x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/TransformIndexer.java

```diff
@@ -508,6 +522,100 @@ protected void onAbort() {
         context.shutdown();
     }
 
+    /**
+     * Let the indexer stop at the next checkpoint and call the listener after the flag has been persisted in state.
+     *
+     * If the indexer isn't running, persist state if required and call the listener immediately.
+     */
+    final void setStopAtCheckpoint(boolean shouldStopAtCheckpoint, ActionListener<Void> shouldStopAtCheckpointListener) {
+        // this should be called from the generic threadpool
+        assert Thread.currentThread().getName().contains(ThreadPool.Names.GENERIC);
```

Comment:
In testing (or where assertions are enabled), this would cause the caller to freeze and never return.

I am not sure how to assert here AND fire an `onFailure` result. But, not doing so might cause some frustrating test investigations